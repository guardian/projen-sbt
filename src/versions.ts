// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
export const minNodeVersion = '14.17.3';
export const maxNodeVersion = '14.17.3';
