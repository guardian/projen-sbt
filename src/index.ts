import * as pj from 'projen';
import { FileBase } from 'projen';
import { Projenrc, ProjenrcOptions } from 'projen/lib/javascript';
import { ProjectOptions } from 'projen/lib/project';
import { maxNodeVersion } from './versions';

export interface SbtPlugin {
  // TODO: for now this is a string but we could type it more nicely
  readonly line: string;
}

export interface SbtProjectOptions extends ProjectOptions {
  readonly sbtVersion?: string;
  readonly projectVersion?: string;
  readonly scalaVersion?: string;

  readonly sbtPlugins?: SbtPlugin[];

  readonly libraryDependencies?: SbtLibraryDependency[];

  readonly projenrcJs?: boolean;
  readonly projenrcJsOptions?: ProjenrcOptions;
}

const defaultSbtVersion = '1.5.5';
const defaultProjectVersion = '1.0-SNAPSHOT';
const defaultScalaVersion = '2.13.6';

const gitIgnored = [
  '**/target/',
  '.idea',
  '.bsp',
];

export type SbtLibraryDependency = string;

export class SbtProject extends pj.Project {
  public static javaDep(org: string, name: string, revision: string, configurations?: string) {
    return SbtProject.dep(org, '%', name, revision, configurations);
  }

  public static scalaDep(org: string, name: string, revision: string, configurations?: string) {
    return SbtProject.dep(org, '%%', name, revision, configurations);
  }

  private static dep(org: string, depSymbol: string, name: string, revision: string, configurations?: string): SbtLibraryDependency {
    const maybeConfigurations = configurations ? ` % "${configurations}"` : '';
    return `"${org}" ${depSymbol} "${name}" % "${revision}"${maybeConfigurations}`;
  }

  readonly sbtPluginsFile: pj.SourceCode;
  readonly buildSbtFile: pj.SourceCode;

  constructor(options: SbtProjectOptions) {
    super(options);

    const sbtVersion = options.sbtVersion ?? defaultSbtVersion;
    const projectVersion = options.projectVersion ?? defaultProjectVersion;
    const scalaVersion = options.scalaVersion ?? defaultScalaVersion;

    const buildPropertiesFile = new pj.SourceCode(this, 'project/build.properties');
    buildPropertiesFile.line(`# ${FileBase.PROJEN_MARKER}`);
    buildPropertiesFile.line(`sbt.version=${sbtVersion}`);

    this.buildSbtFile = new pj.SourceCode(this, 'build.sbt');
    this.buildSbtFile.line(`// ${FileBase.PROJEN_MARKER}`);
    this.buildSbtFile.line(`name := "${options.name}"`);
    this.buildSbtFile.line(`version := "${projectVersion}"`);
    this.buildSbtFile.line(`scalaVersion in ThisBuild := "${scalaVersion}"`);

    options.libraryDependencies?.forEach(dep => {
      this.buildSbtFile.line(`libraryDependencies += ${dep}`);
    });

    this.sbtPluginsFile = new pj.SourceCode(this, 'project/plugins.sbt');
    this.sbtPluginsFile.line(`// ${FileBase.PROJEN_MARKER}`);
    if (options.sbtPlugins) {
      this.addPlugins(...options.sbtPlugins);
    }

    this.addTask('compile', {
      exec: 'sbt compile',
    });

    this.addTask('build', {
      exec: 'sbt clean compile test',
    });

    // generate a projen rc file (without this none is generated)
    const projenrcJs = options.projenrcJs ?? true;
    if (projenrcJs) {
      new Projenrc(this, options.projenrcJsOptions);
    }

    const ciWorkflow = this.github?.addWorkflow('Build');

    ciWorkflow?.on({
      pullRequest: {},
      workflowDispatch: {},
    });

    ciWorkflow?.addJobs({
      Build: {
        runsOn: 'ubuntu-latest',
        steps: [
          { uses: 'actions/checkout@v2' },
          {
            uses: 'actions/setup-node@v2.1.5',
            with: { 'node-version': maxNodeVersion },
          },
          // from https://github.com/actions/cache/blob/master/examples.md#scala---sbt
          // TODO: fix to include coursier
          {
            uses: 'actions/cache@v2',
            with: {
              path: ['~/.ivy2/cache', '~/.sbt'].join('\n'),
              key: "${{ runner.os }}-sbt-${{ hashFiles('**/build.sbt') }}",
            },
          },
          {
            uses: 'actions/setup-java@v2',
            with: { 'java-version': '8', 'distribution': 'adopt' },
          },
          {
            run: 'yarn',
          },
          {
            run: 'npx projen build',
          },
        ],
        permissions: {},
      },
    });


    gitIgnored.forEach(pattern => this.gitignore.exclude(pattern));
  }

  public addPlugins(...plugins: SbtPlugin[]) {
    plugins.forEach(plugin => {
      this.sbtPluginsFile.line(plugin.line);
    });
  }
}
