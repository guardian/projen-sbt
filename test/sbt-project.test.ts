import { LogLevel } from 'projen';
import { synthSnapshot, mkdtemp, SynthOutput } from 'projen/lib/__tests__/util';
import { SbtProject } from '../src';

describe('The SbtProject type', () => {
  const baseOptions = {
    name: 'test',
    outdir: mkdtemp(),
    logging: {
      level: LogLevel.OFF,
    },
  };

  test('a simple SbtProject', () => {
    const sbtProject = new SbtProject(baseOptions);

    const snapshot = synthSnapshot(sbtProject) as SynthOutput;
    expect(snapshot).toMatchSnapshot();
  });

  test('overriding the sbt version', () => {
    const sbtProject = new SbtProject({
      ...baseOptions,
      sbtVersion: 'x.y.z',
    });
    const snapshot = synthSnapshot(sbtProject) as SynthOutput;
    expect(snapshot['project/build.properties']).toMatchSnapshot();
  });

  test('overriding the project version', () => {
    const sbtProject = new SbtProject({
      ...baseOptions,
      projectVersion: 'x.y.z',
    });
    const snapshot = synthSnapshot(sbtProject) as SynthOutput;
    expect(snapshot['build.sbt']).toMatchSnapshot();
  });

  test('overriding the scala version', () => {
    const sbtProject = new SbtProject({
      ...baseOptions,
      scalaVersion: 'x.y.z',
    });
    const snapshot = synthSnapshot(sbtProject) as SynthOutput;
    expect(snapshot['build.sbt']).toMatchSnapshot();
  });

  test('adding sbt plugins', () => {
    const sbtProject = new SbtProject({
      ...baseOptions,
      sbtPlugins: [
        { line: 'addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.8.8")' },
        { line: 'libraryDependencies += "org.vafer" % "jdeb" % "1.8" artifacts Artifact("jdeb", "jar", "jar")' },
      ],
    });
    const snapshot = synthSnapshot(sbtProject) as SynthOutput;
    expect(snapshot['project/plugins.sbt']).toMatchSnapshot();
  });
});
