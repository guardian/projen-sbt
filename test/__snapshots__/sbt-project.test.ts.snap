// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The SbtProject type a simple SbtProject 1`] = `
Object {
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

/.gitattributes             	linguist-generated
/.github/workflows/build.yml	linguist-generated
/.github/workflows/stale.yml	linguist-generated
/.gitignore                 	linguist-generated
/.projen/**                 	linguist-generated
/.projen/deps.json          	linguist-generated
/.projen/tasks.json         	linguist-generated
/build.sbt                  	linguist-generated
/project/build.properties   	linguist-generated
/project/plugins.sbt        	linguist-generated",
  ".github/workflows/build.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: Build
on:
  pull_request: {}
  workflow_dispatch: {}
jobs:
  Build:
    runs-on: ubuntu-latest
    permissions: {}
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v2.1.5
        with:
          node-version: 14.17.3
      - uses: actions/cache@v2
        with:
          path: |-
            ~/.ivy2/cache
            ~/.sbt
          key: \${{ runner.os }}-sbt-\${{ hashFiles('**/build.sbt') }}
      - uses: actions/setup-java@v2
        with:
          java-version: \\"8\\"
          distribution: adopt
      - run: yarn
      - run: npx projen build
",
  ".github/workflows/stale.yml": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: stale
on:
  schedule:
    - cron: 0 1 * * *
  workflow_dispatch: {}
jobs:
  stale:
    runs-on: ubuntu-latest
    permissions:
      issues: write
      pull-requests: write
    steps:
      - uses: actions/stale@v4
        with:
          days-before-stale: -1
          days-before-close: -1
          days-before-pr-stale: 14
          days-before-pr-close: 2
          stale-pr-message: This pull request is now marked as stale because it hasn't
            seen activity for a while. Add a comment or it will be closed soon.
          close-pr-message: Closing this pull request as it hasn't seen activity for a
            while. Please add a comment @mentioning a maintainer to reopen.
          stale-pr-label: stale
          days-before-issue-stale: 60
          days-before-issue-close: 7
          stale-issue-message: This issue is now marked as stale because it hasn't seen
            activity for a while. Add a comment or it will be closed soon.
          close-issue-message: Closing this issue as it hasn't seen activity for a while.
            Please add a comment @mentioning a maintainer to reopen.
          stale-issue-label: stale
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
**/target/
.bsp
.idea
node_modules/
!/.gitattributes
!/.github/workflows/build.yml
!/.github/workflows/stale.yml
!/.projen/deps.json
!/.projen/tasks.json
!/build.sbt
!/project/build.properties
!/project/plugins.sbt
",
  ".projen/tasks.json": Object {
    "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
    "tasks": Object {
      "build": Object {
        "name": "build",
        "steps": Array [
          Object {
            "exec": "sbt clean compile test",
          },
        ],
      },
      "clobber": Object {
        "condition": "git diff --exit-code > /dev/null",
        "description": "hard resets to HEAD of origin and cleans the local repo",
        "env": Object {
          "BRANCH": "$(git branch --show-current)",
        },
        "name": "clobber",
        "steps": Array [
          Object {
            "exec": "git checkout -b scratch",
            "name": "save current HEAD in \\"scratch\\" branch",
          },
          Object {
            "exec": "git checkout $BRANCH",
          },
          Object {
            "exec": "git fetch origin",
            "name": "fetch latest changes from origin",
          },
          Object {
            "exec": "git reset --hard origin/$BRANCH",
            "name": "hard reset to origin commit",
          },
          Object {
            "exec": "git clean -fdx",
            "name": "clean all untracked files",
          },
          Object {
            "say": "ready to rock! (unpushed commits are under the \\"scratch\\" branch)",
          },
        ],
      },
      "compile": Object {
        "name": "compile",
        "steps": Array [
          Object {
            "exec": "sbt compile",
          },
        ],
      },
      "default": Object {
        "name": "default",
        "steps": Array [
          Object {
            "exec": "node .projenrc.js",
          },
        ],
      },
    },
  },
  "README.md": "# replace this",
  "build.sbt": "// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
name := \\"test\\"
version := \\"1.0-SNAPSHOT\\"
scalaVersion in ThisBuild := \\"2.13.6\\"",
  "project/build.properties": "# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
sbt.version=1.5.5",
  "project/plugins.sbt": "// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
}
`;

exports[`The SbtProject type adding library dependencies 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
name := \\"test\\"
version := \\"1.0-SNAPSHOT\\"
scalaVersion in ThisBuild := \\"2.13.6\\"
libraryDependencies += \\"joda-time\\" % \\"joda-time\\" % \\"2.8.1\\"
libraryDependencies += \\"org.scalamock\\" %% \\"scalamock\\" % \\"4.0.0\\" % \\"test\\""
`;

exports[`The SbtProject type adding sbt plugins 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
addSbtPlugin(\\"com.typesafe.play\\" % \\"sbt-plugin\\" % \\"2.8.8\\")
libraryDependencies += \\"org.vafer\\" % \\"jdeb\\" % \\"1.8\\" artifacts Artifact(\\"jdeb\\", \\"jar\\", \\"jar\\")"
`;

exports[`The SbtProject type overriding the project version 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
name := \\"test\\"
version := \\"x.y.z\\"
scalaVersion in ThisBuild := \\"2.13.6\\""
`;

exports[`The SbtProject type overriding the sbt version 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
sbt.version=x.y.z"
`;

exports[`The SbtProject type overriding the scala version 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".
name := \\"test\\"
version := \\"1.0-SNAPSHOT\\"
scalaVersion in ThisBuild := \\"x.y.z\\""
`;
